File: services/finance/engines/ratios_engine.py
from pydantic import BaseModel
from typing import Optional


class FinancialStatements(BaseModel):
    current_assets: float
    current_liabilities: float
    cash: float
    total_assets: float
    total_equity: float
    total_liabilities: float
    ebit: float
    net_income: float
    revenue: float
    ebitda: float
    inventories: float
    accounts_receivable: float
    accounts_payable: float
    cost_of_goods_sold: float
    interest_expense: float
    operating_cashflow: float


class LiquidityRatios(BaseModel):
    current_ratio: float
    quick_ratio: float
    cash_ratio: float
    working_capital: float


class ProfitabilityRatios(BaseModel):
    gross_margin: float
    operating_margin: float
    net_margin: float
    ebitda_margin: float
    roa: float
    roe: float
    roce: float


class LeverageRatios(BaseModel):
    debt_to_equity: float
    debt_ratio: float
    equity_ratio: float
    interest_coverage: float
    debt_to_ebitda: float


class EfficiencyRatios(BaseModel):
    asset_turnover: float
    inventory_turnover: float
    receivables_turnover: float
    payables_turnover: float
    days_sales_outstanding: float
    days_inventory_outstanding: float
    days_payables_outstanding: float
    cash_conversion_cycle: float


class RatiosEngine:
    def liquidity(self, fs: FinancialStatements) -> LiquidityRatios:
        cr = fs.current_assets / fs.current_liabilities if fs.current_liabilities != 0 else 0.0
        qr = (fs.current_assets - fs.inventories) / fs.current_liabilities if fs.current_liabilities != 0 else 0.0
        cash_r = fs.cash / fs.current_liabilities if fs.current_liabilities != 0 else 0.0
        wc = fs.current_assets - fs.current_liabilities
        return LiquidityRatios(
            current_ratio=round(cr, 2),
            quick_ratio=round(qr, 2),
            cash_ratio=round(cash_r, 2),
            working_capital=round(wc, 2),
        )

    def profitability(self, fs: FinancialStatements) -> ProfitabilityRatios:
        gm = ((fs.revenue - fs.cost_of_goods_sold) / fs.revenue * 100) if fs.revenue != 0 else 0.0
        om = (fs.ebit / fs.revenue * 100) if fs.revenue != 0 else 0.0
        nm = (fs.net_income / fs.revenue * 100) if fs.revenue != 0 else 0.0
        em = (fs.ebitda / fs.revenue * 100) if fs.revenue != 0 else 0.0
        roa = (fs.net_income / fs.total_assets * 100) if fs.total_assets != 0 else 0.0
        roe = (fs.net_income / fs.total_equity * 100) if fs.total_equity != 0 else 0.0
        ce = fs.total_assets - fs.total_liabilities
        roce = (fs.ebit / ce * 100) if ce != 0 else 0.0
        return ProfitabilityRatios(
            gross_margin=round(gm, 2),
            operating_margin=round(om, 2),
            net_margin=round(nm, 2),
            ebitda_margin=round(em, 2),
            roa=round(roa, 2),
            roe=round(roe, 2),
            roce=round(roce, 2),
        )

    def leverage(self, fs: FinancialStatements) -> LeverageRatios:
        de = fs.total_liabilities / fs.total_equity if fs.total_equity != 0 else 0.0
        dr = fs.total_liabilities / fs.total_assets if fs.total_assets != 0 else 0.0
        er = fs.total_equity / fs.total_assets if fs.total_assets != 0 else 0.0
        ic = fs.ebit / fs.interest_expense if fs.interest_expense != 0 else 0.0
        d_ebitda = fs.total_liabilities / fs.ebitda if fs.ebitda != 0 else 0.0
        return LeverageRatios(
            debt_to_equity=round(de, 2),
            debt_ratio=round(dr, 2),
            equity_ratio=round(er, 2),
            interest_coverage=round(ic, 2),
            debt_to_ebitda=round(d_ebitda, 2),
        )

    def efficiency(self, fs: FinancialStatements) -> EfficiencyRatios:
        at = fs.revenue / fs.total_assets if fs.total_assets != 0 else 0.0
        it = fs.cost_of_goods_sold / fs.inventories if fs.inventories != 0 else 0.0
        rt = fs.revenue / fs.accounts_receivable if fs.accounts_receivable != 0 else 0.0
        pt = fs.cost_of_goods_sold / fs.accounts_payable if fs.accounts_payable != 0 else 0.0
        dso = 365 / rt if rt != 0 else 0.0
        dio = 365 / it if it != 0 else 0.0
        dpo = 365 / pt if pt != 0 else 0.0
        ccc = dso + dio - dpo
        return EfficiencyRatios(
            asset_turnover=round(at, 2),
            inventory_turnover=round(it, 2),
            receivables_turnover=round(rt, 2),
            payables_turnover=round(pt, 2),
            days_sales_outstanding=round(dso, 2),
            days_inventory_outstanding=round(dio, 2),
            days_payables_outstanding=round(dpo, 2),
            cash_conversion_cycle=round(ccc, 2),
        )

    def all_ratios(self, fs: FinancialStatements) -> dict:
        return {
            "liquidity": self.liquidity(fs).model_dump(),
            "profitability": self.profitability(fs).model_dump(),
            "leverage": self.leverage(fs).model_dump(),
            "efficiency": self.efficiency(fs).model_dump(),
        }

File: cos/services/finance/src/finance/ratios-engine.ts
import { Injectable } from "@nestjs/common"

export interface FinancialInput {
  currentAssets: number
  currentLiabilities: number
  totalAssets: number
  totalLiabilities: number
  equity: number
  revenue: number
  netIncome: number
  ebit: number
  inventory: number
  accountsReceivable: number
  cogs: number
  ebitda: number
  interestExpense: number
  cash: number
  accountsPayable: number
  workingCapital: number
}

export interface RiskSignal {
  metric: string
  severity: "low" | "medium" | "high" | "critical"
  message: string
  value: number
  threshold: number
}

@Injectable()
export class RatiosEngine {
  calculate(data: FinancialInput, prevData?: FinancialInput) {
    const ratios = {
      liquidity: {
        currentRatio: data.currentLiabilities ? +(data.currentAssets / data.currentLiabilities).toFixed(2) : null,
        quickRatio: data.currentLiabilities ? +((data.currentAssets - data.inventory) / data.currentLiabilities).toFixed(2) : null,
        cashRatio: data.currentLiabilities ? +(data.cash / data.currentLiabilities).toFixed(2) : null,
        workingCapitalRatio: data.totalAssets ? +(data.workingCapital / data.totalAssets).toFixed(2) : null,
      },
      solvency: {
        debtToEquity: data.equity ? +(data.totalLiabilities / data.equity).toFixed(2) : null,
        debtRatio: data.totalAssets ? +(data.totalLiabilities / data.totalAssets).toFixed(2) : null,
        equityRatio: data.totalAssets ? +((data.equity / data.totalAssets) * 100).toFixed(2) : null,
        interestCoverage: data.interestExpense ? +(data.ebit / data.interestExpense).toFixed(2) : null,
      },
      profitability: {
        netMargin: data.revenue ? +((data.netIncome / data.revenue) * 100).toFixed(2) : null,
        operatingMargin: data.revenue ? +((data.ebit / data.revenue) * 100).toFixed(2) : null,
        roa: data.totalAssets ? +((data.netIncome / data.totalAssets) * 100).toFixed(2) : null,
        roe: data.equity ? +((data.netIncome / data.equity) * 100).toFixed(2) : null,
        ebitdaMargin: data.revenue ? +((data.ebitda / data.revenue) * 100).toFixed(2) : null,
      },
      efficiency: {
        assetTurnover: data.totalAssets ? +(data.revenue / data.totalAssets).toFixed(2) : null,
        inventoryTurnover: data.inventory ? +(data.cogs / data.inventory).toFixed(2) : null,
        receivablesTurnover: data.accountsReceivable ? +(data.revenue / data.accountsReceivable).toFixed(2) : null,
        payablesTurnover: data.accountsPayable ? +(data.cogs / data.accountsPayable).toFixed(2) : null,
      },
    }

    const signals = this.detectRiskSignals(data, ratios)
    const change = prevData ? this.calculateChange(ratios, prevData) : null

    return { ratios, signals, change }
  }

  private detectRiskSignals(data: FinancialInput, ratios: any): RiskSignal[] {
    const signals: RiskSignal[] = []

    if (ratios.liquidity.currentRatio !== null && ratios.liquidity.currentRatio < 1.0) {
      signals.push({ metric: "current_ratio", severity: "critical", message: "Capital de trabajo negativo — riesgo de liquidez inmediato", value: ratios.liquidity.currentRatio, threshold: 1.0 })
    } else if (ratios.liquidity.currentRatio !== null && ratios.liquidity.currentRatio < 1.5) {
      signals.push({ metric: "current_ratio", severity: "high", message: "Liquidez por debajo del nivel recomendado", value: ratios.liquidity.currentRatio, threshold: 1.5 })
    }

    if (ratios.solvency.debtToEquity !== null && ratios.solvency.debtToEquity > 3.0) {
      signals.push({ metric: "debt_to_equity", severity: "critical", message: "Apalancamiento excesivo — riesgo de insolvencia", value: ratios.solvency.debtToEquity, threshold: 3.0 })
    } else if (ratios.solvency.debtToEquity !== null && ratios.solvency.debtToEquity > 2.0) {
      signals.push({ metric: "debt_to_equity", severity: "high", message: "Apalancamiento elevado", value: ratios.solvency.debtToEquity, threshold: 2.0 })
    }

    if (ratios.solvency.interestCoverage !== null && ratios.solvency.interestCoverage < 1.5) {
      signals.push({ metric: "interest_coverage", severity: "critical", message: "EBIT insuficiente para cubrir intereses", value: ratios.solvency.interestCoverage, threshold: 1.5 })
    } else if (ratios.solvency.interestCoverage !== null && ratios.solvency.interestCoverage < 3.0) {
      signals.push({ metric: "interest_coverage", severity: "high", message: "Cobertura de intereses ajustada", value: ratios.solvency.interestCoverage, threshold: 3.0 })
    }

    if (ratios.profitability.netMargin !== null && ratios.profitability.netMargin < 0) {
      signals.push({ metric: "net_margin", severity: "critical", message: "Margen neto negativo — pérdidas operativas", value: ratios.profitability.netMargin, threshold: 0 })
    }

    if (ratios.profitability.roe !== null && ratios.profitability.roe < 5) {
      signals.push({ metric: "roe", severity: "medium", message: "Rendimiento sobre capital por debajo del costo de oportunidad", value: ratios.profitability.roe, threshold: 5 })
    }

    if (ratios.liquidity.cashRatio !== null && ratios.liquidity.cashRatio < 0.2) {
      signals.push({ metric: "cash_ratio", severity: "high", message: "Efectivo insuficiente para obligaciones inmediatas", value: ratios.liquidity.cashRatio, threshold: 0.2 })
    }

    if (ratios.efficiency.receivablesTurnover !== null && ratios.efficiency.receivablesTurnover < 4) {
      signals.push({ metric: "receivables_turnover", severity: "medium", message: "Cobranza lenta — posible riesgo de incobrabilidad", value: ratios.efficiency.receivablesTurnover, threshold: 4 })
    }

    return signals
  }

  private calculateChange(ratios: any, prev: FinancialInput) {
    const prevCurrentRatio = prev.currentLiabilities ? prev.currentAssets / prev.currentLiabilities : null
    const prevNetMargin = prev.revenue ? (prev.netIncome / prev.revenue) * 100 : null
    const prevRoa = prev.totalAssets ? (prev.netIncome / prev.totalAssets) * 100 : null
    const prevRoe = prev.equity ? (prev.netIncome / prev.equity) * 100 : null
    return {
      currentRatio: prevCurrentRatio ? +((ratios.liquidity.currentRatio - prevCurrentRatio) / prevCurrentRatio * 100).toFixed(1) : null,
      netMargin: prevNetMargin ? +((ratios.profitability.netMargin - prevNetMargin) / prevNetMargin * 100).toFixed(1) : null,
      roa: prevRoa ? +((ratios.profitability.roa - prevRoa) / prevRoa * 100).toFixed(1) : null,
      roe: prevRoe ? +((ratios.profitability.roe - prevRoe) / prevRoe * 100).toFixed(1) : null,
    }
  }
}
